src/windows/video_widget.py

```diff
@@ -509,15 +509,23 @@ def paintEvent(self, event, *args):
     def centeredViewport(self, width, height):
         """ Calculate size of viewport to maintain aspect ratio """
 
-        window_size = QSizeF(width, height)
-        window_rect = QRectF(QPointF(0, 0), window_size)
+        # Calculate padding
+        top_padding = (height - (height * self.zoom)) / 2.0
+        left_padding = (width - (width * self.zoom)) / 2.0
 
-        aspectRatio = self.aspect_ratio.ToFloat() * self.pixel_ratio.ToFloat()
-        viewport_size = QSizeF(aspectRatio, 1).scaled(window_size, Qt.KeepAspectRatio)
-        viewport_rect = QRectF(QPointF(0, 0), viewport_size)
-        viewport_rect.moveCenter(window_rect.center())
+        # Adjust parameters to zoom
+        width = width * self.zoom
+        height = height * self.zoom
 
-        return viewport_rect.toRect()
+        # Calculate which direction to scale (for perfect centering)
+        aspectRatio = self.aspect_ratio.ToFloat()
+        heightFromWidth = width / aspectRatio
+        widthFromHeight = height * aspectRatio
+
+        if heightFromWidth <= height:
+            return QRect(left_padding, ((height - heightFromWidth) / 2) + top_padding, width, heightFromWidth)
+        else:
+            return QRect(((width - widthFromHeight) / 2.0) + left_padding, top_padding, widthFromHeight, height)
```

Comment:
This change re-broke Python 3.10. you can't pass `QRect()` float values, it only takes `int`s. In Python 3.10 that's an instant crash.

If there's an issue here, it's that my code doesn't account for zoom — I _thought_ that was intentional, I thought I'd handled that somewhere else. I'd certainly _meant_ to, because the zoom math should be independent of the viewport. The viewport math is based on the original, un-zoomed frame image.

If it gets zoomed in or out, that just scales it larger or smaller around its own center point. Zooming changes the _apparent_ size, not the _actual_ size. If the viewport is going to account for zoom, then shouldn't it have to  account for rotation as well? Rotation also changes the frame's apparent size (and aspect).

Reply:
(The idea here was that `centeredViewport()` would be generic — you give it dimensions, it tells you how to center the image in those dimensions based on its aspect ratio. Scaling, rotation, etc, are all separate operations that occur after that point.)

Reply:
Also, at the risk of ranting a bit... the code needs to stop abusing the term "heightForWidth" for simple scaling. That term has a meaning in Qt, and that's not what it means.

"HeightForWidth" means that a widget's height is dependent on its width, **INDEPENDENT of scaling**.

Picture a text widget containing some wrapped text. If the widget is wide enough that none of the lines wrap, N lines fit in a widget N lines tall:

![image](https://user-images.githubusercontent.com/538020/147868098-76712f94-3872-4d20-904a-9893b43b6a12.png)

But if the widget is made narrower and lines wrap, its height changes:

![image](https://user-images.githubusercontent.com/538020/147868111-9ff448e3-dd12-48da-9f50-a66c3dd51d7a.png)

**THAT** widget would return `True` for `.hasHeightForWidth()`. Because that's what it means. Scaling a rectangle is not a height-for-width operation, it's just scaling. And it doesn't need to be done in the complicated way it's done here.